async-nats/src/lib.rs

```diff
@@ -501,9 +509,33 @@ impl Connector {
     pub(crate) async fn try_connect(&mut self) -> Result<(ServerInfo, Connection), io::Error> {
         let mut error = None;
 
-        for server_addr in &self.server_addrs {
-            match self.try_connect_to(server_addr).await {
-                Ok(inner) => return Ok(inner),
+        let server_addrs: Vec<ServerAddr> = self.attempts.keys().cloned().collect();
+        for server_addr in server_addrs {
+            let server_attempts = self.attempts.get_mut(&server_addr).unwrap();
+            let duration = if *server_attempts == 0 {
+                Duration::from_millis(0)
+            } else {
+                let exp: u32 = (*server_attempts - 1).try_into().unwrap_or(std::u32::MAX);
+                let max = Duration::from_secs(4);
+
+                cmp::min(Duration::from_millis(2_u64.saturating_pow(exp)), max)
+            };
+
+            *server_attempts += 1;
+            sleep(duration).await;
+
+            match self.try_connect_to(&server_addr).await {
+                Ok((info, connection)) => {
+                    for url in &info.connect_urls {
+                        self.attempts.insert(url.parse()?, 0);
```

Comment:
This will reset attempts for any erroneous server that is in the info list (and those URLs are prone to errors as the may be internal and not accessible from the Client)

Reply:
Semantics yanked from nats/connector.rs

Reply:
Not really. You got all advertised URLs here, not only the one that actually works. Is fine to reset only the one we connected to.